nixos/modules/tasks/auto-upgrade.nix

```
@@ -189,6 +494,8 @@ in
       }
     ];
 
+    environment.systemPackages = lib.optional (cfg.desync != { }) upgradeDesyncsPkg;
```

Probably better to use `mkIf` than `optional`. Unless this was for a specific reason such as avoiding inf-rec.

```suggestion
    environment.systemPackages = lib.mkIf (cfg.desync != { }) [
      upgradeDesyncsPkg
    ];
```

---

I'm not sure I see the advantage of using `mkIf` here?

---

`optional false` results in a definition of `[ ]` while `mkIf false` (usually\*) results in no definition.

Practically speaking, there's not _much_ difference to the final value. But there is a subtle difference in how the definitions get merged, how they'd appear in warnings, etc.

\* one exception being "lazy" types that have an `emptyValue` used when undefined

---

That's actually why I prefer `optional` whenever possible: IIRC `mkIf false` results in something like `{ if = false; value = <thunk>; }`, which I'd expect would slow down eval. While otherwise there would be basically no practical impact.

Does that make sense, or has the module system changed too much since I last looked at it?

---

You are correct, `mkIf` will wrap the value. However imo it's better to try to use the module system idiomatically, rather than attempt to micro-optimise these kinda things.

As for which is actually more optimal, I'm unsure. On the one hand `mkIf false` will result in one less definition being passed to the type's merge function. On the other hand, merging an empty list as one of the definitions shouldn't be expensive.

---

One scenario where `mkIf false` is better than an unconditional "empty" definition is when dealing with multiple override priorities.

E.g. a user defines `mkDefault [ "foo" ]` and some other module defines `mkIf false [ "bar" ]`. Due to using `mkIf`, the second definition is skipped and the `highestPrio` is `(mkDefault null).priority` (1000). Therefore the final value is `[ "foo" ]`.

If we tweak that example to use `if then else` or `optionals` instead of `mkIf`, then the second definition is now unconditional\* and will always override the `mkDefault` definition. Therefore the `highestPrio` will be `defaultOverridePriority` (100) and the final value will be `[ ]`.

\* the definition's _value_ is conditional, but the definition itself is not.

---

Although, in this specific scenario, it seems almost impossible that `environment.systemPackages` would ever end up with a lower priority than `defaultOverridePriority`, given it is defined by the majority of modules.